### PR TITLE
fix(#2444): useRestoreFocus when nodeToRestore has been unmounted

### LINKED ISSUE #2444: FocusScope fails to restore focus when the element to restore has unmounted
# 🐛 Bug Report 

In the [DialogContainer: Dialog triggered by a menu item](https://react-spectrum.adobe.com/react-spectrum/DialogContainer.html#dialog-triggered-by-a-menu-item) example, a Dialog opens from a menu item, and then the menu containing the menu item closes, removing the element to which focus should be restored when the Dialog closes from the DOM. When the Dialog closes, focus is lost to the `document.body`. 

## 🤔 Expected Behavior

When a scope containing a `nodeToRestore` has been unmounted, we should restore focus to a previous `nodeToRestore, so in the [Dialog triggered by a menu item](https://react-spectrum.adobe.com/react-spectrum/DialogContainer.html#dialog-triggered-by-a-menu-item)  example, focus should return to the ActionButton within the MenuTrigger.

## 😯 Current Behavior

When the Dialog closes, focus is lost to the `document.body`, because we do not keep track of previous `nodeToRestore`s.

## 💁 Possible Solution

Manage a stack of `nodeToRestore` elements so that when unmounting a scope, if that scopes `nodeToRestore` is no longer in the DOM, we can restore focus to previous `nodeToRestore` that remains in the DOM.

## 🔦 Context

This is very important for keyboard and screen reader accessibility, especially with Portals in React, because when a FocusScope within a Portal unmounts with nowhere to return focus, focus is lost to the `document.body` at the end of the DOM, and a keyboard or screen reader user would have to navigate through the document to get back to where they left off.

A user can explicitly manage focus within their application, when dialogs close, but that puts the burden on the developer, when it should probably be the framework's responsibility.

## 💻 Code Sample

Within FocusScope: 

```ts
// An array to create a stack of nodeToRestore elements.
let nodeToRestoreArray: HTMLElement[] = [];

function addToNodeToRestoreArray(node: HTMLElement) {
  if (nodeToRestoreArray.indexOf(node) === -1) {
    nodeToRestoreArray.push(node);
  }
}

function removeFromNodeToRestoreArray(node: HTMLElement) {
  let index = nodeToRestoreArray.indexOf(node);
  if (index > -1) {
    nodeToRestoreArray.splice(index, 1);
  }
}

function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boolean, contain: boolean) {
  // useLayoutEffect instead of useEffect so the active element is saved synchronously instead of asynchronously.
  useLayoutEffect(() => {
    if (!restoreFocus) {
      return;
    }

    let scope = scopeRef.current;
    let nodeToRestore = document.activeElement as HTMLElement;
    addToNodeToRestoreArray(nodeToRestore);

    // Handle the Tab key so that tabbing out of the scope goes to the next element
    // after the node that had focus when the scope mounted. This is important when
    // using portals for overlays, so that focus goes to the expected element when
    // tabbing out of the overlay.
    let onKeyDown = (e: KeyboardEvent) => {
      if (e.key !== 'Tab' || e.altKey || e.ctrlKey || e.metaKey) {
        return;
      }

      let focusedElement = document.activeElement as HTMLElement;
      if (!isElementInScope(focusedElement, scope)) {
        return;
      }

      // Create a DOM tree walker that matches all tabbable elements
      let walker = getFocusableTreeWalker(document.body, {tabbable: true});

      // Find the next tabbable element after the currently focused element
      walker.currentNode = focusedElement;
      let nextElement = (e.shiftKey ? walker.previousNode() : walker.nextNode()) as HTMLElement;

      if (!document.body.contains(nodeToRestore) || nodeToRestore === document.body) {
        removeFromNodeToRestoreArray(nodeToRestore);
        nodeToRestore = null;
      }

      // If there is no next element, or it is outside the current scope, move focus to the
      // next element after the node to restore to instead.
      if ((!nextElement || !isElementInScope(nextElement, scope)) && nodeToRestore) {
        walker.currentNode = nodeToRestore;

        // Skip over elements within the scope, in case the scope immediately follows the node to restore.
        do {
          nextElement = (e.shiftKey ? walker.previousNode() : walker.nextNode()) as HTMLElement;
        } while (isElementInScope(nextElement, scope));

        e.preventDefault();
        e.stopPropagation();
        if (nextElement) {
          focusElement(nextElement, true);
        } else {
           // If there is no next element and the nodeToRestore isn't within a FocusScope (i.e. we are leaving the top level focus scope)
           // then move focus to the body.
           // Otherwise restore focus to the nodeToRestore (e.g menu within a popover -> tabbing to close the menu should move focus to menu trigger)
          if (!isElementInAnyScope(nodeToRestore)) {
            focusedElement.blur();
          } else {
            focusElement(nodeToRestore, true);
          }
        }
      }
    };

    if (!contain) {
      document.addEventListener('keydown', onKeyDown, true);
    }

    return () => {
      if (!contain) {
        document.removeEventListener('keydown', onKeyDown, true);
      }

      if (nodeToRestore) {
        let index = nodeToRestoreArray.indexOf(nodeToRestore);

        if (restoreFocus && isElementInScope(document.activeElement, scope)) {
          requestAnimationFrame(() => {
            if (document.body.contains(nodeToRestore) && nodeToRestore !== document.body) {
              focusElement(nodeToRestore);
            } else {
              while (index > 0) {
                index--;
                let node = nodeToRestoreArray[index];
                if (!document.body.contains(node) || node === document.body) {
                  removeFromNodeToRestoreArray(node);
                } else {
                  focusElement(node);
                  break;
                }
              }
            }
          });
        } else if (!document.body.contains(nodeToRestore)) {
          removeFromNodeToRestoreArray(nodeToRestore);
        }
      }
    };
  }, [scopeRef, restoreFocus, contain]);
}
```

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  @react-aria/focus v3.5.0
| Browser          | All
| Operating System | All

## 🧢 Your Company/Team

Adobe/Accessibility

## 🕷 Tracking Issue (optional)

Related issues: 

- https://github.com/adobe/react-spectrum/issues/874#issuecomment-816127130
- https://github.com/adobe/react-spectrum/issues/1130
- https://github.com/adobe/react-spectrum/issues/2415
- https://github.com/adobe/react-spectrum/issues/2421
- https://github.com/adobe/react-spectrum/pull/2416

### PR DESCRIPTION
**Pull request by majornista**
Restore focus to previous nodeToRestore when expected nodeToRestore is no longer present, as with a Menu that opens a Dialog.

Closes #2444

## ✅ Pull Request Checklist:

- [X] Included link to corresponding React Spectrum GitHub Issue #2444.
- [X] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [X] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/4b09194340f607fab1c56739c71f99e8a7d58337/storybook/index.html?path=/story/dialogcontainer--in-a-menu
2. Click or navigate to the "Open menu" button and press Enter or Space to expand the menu.
3. Click or arrow down to and activate the "Open dialog..." menu item.
4. Press Esc key to close the Dialog.
5. Focus should be restored to the "Open menu" button.

1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/4b09194340f607fab1c56739c71f99e8a7d58337/storybook/index.html?path=/story/dialogcontainer--nested-dialog-containers
2. Click or navigate to the "Open menu" button and press Enter or Space to expand the menu.
3. Click or arrow down to and activate the "Do this..." or "Do that..." menu item.
4. Click or activate the "Do that" or "Do this" button to replace the Dialog with the nested Dialog.
5. Repeat step 4 a few times for good measure.
6. Press Esc key, or use the "Dismiss" button to close the Dialog.
7. Focus should be restored to the "Open menu" button.

Verify that https://reactspectrum.blob.core.windows.net/reactspectrum/4b09194340f607fab1c56739c71f99e8a7d58337/storybook/index.html?path=/story/focusscope--keyboard-navigation and https://reactspectrum.blob.core.windows.net/reactspectrum/4b09194340f607fab1c56739c71f99e8a7d58337/storybook/index.html?path=/story/focusscope--keyboard-navigation-inside-portal manage and restore focus appropriately.

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 1]
```diff
@@ -383,20 +385,45 @@ function useAutoFocus(scopeRef: RefObject<HTMLElement[]>, autoFocus: boolean) {
...
+    let nodeToRestore: HTMLElement;
+
+    if (typeof restoreFocus === 'boolean') {
+      nodeToRestore = nodeToRestoreRef.current as HTMLElement;
+    } else if (restoreFocus.current) {
+      nodeToRestore = restoreFocus.current instanceof HTMLElement ? restoreFocus.current : restoreFocus.current.UNSAFE_getDOMNode();
+    }
+
+    addToNodeToRestoreArray(nodeToRestore);
+
```
Comment (LFDanLu) [inline packages/@react-aria/focus/src/FocusScope.tsx:426]: IMO we should still save the `document.activeElement` in the `nodeToRestoreArray` even if the user provides a node to `restoreFocus` to cover for a case where the supplied `restoreFocus` node is removed before the FocusScope is unmounted.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 2]
```diff
@@ -458,12 +486,29 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
...
+                }
+              }
+            }
+          });
+        } else if (!document.body.contains(nodeToRestore)) {
+          removeFromNodeToRestoreArray(nodeToRestore);
+        }
       }
     };
   }, [scopeRef, restoreFocus, contain]);
```
Comment (LFDanLu) [inline packages/@react-aria/focus/src/FocusScope.tsx:514]: As an aside and probably out of scope of this PR since it is preexisting, but doesn't this `return` fire not only as a cleanup but also if any of these dep array values change? That means we could unintentionally call `focusElement` and move focus out of the active scope if `contain` were to update values

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 1 above)
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:426]: That make sense.

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx [hunk 3]
```diff
@@ -116,3 +123,55 @@ KeyboardNavigationNoContain.args = {usePortal: false, contain: false};
 
 export const KeyboardNavigationInsidePortalNoContain = Template().bind({});
 KeyboardNavigationInsidePortalNoContain.args = {usePortal: true, contain: false};
+
+export const RestoreFocusRef = Template().bind({});
+RestoreFocusRef.args = {usePortal: false, useRestoreFocusRef: true};
+
+export const RestoreFocusRefInsidePortal = Template().bind({});
+RestoreFocusRefInsidePortal.args = {usePortal: true, useRestoreFocusRef: true};
```
Comment (LFDanLu) [inline packages/@react-aria/focus/stories/FocusScope.stories.tsx:131]: Strange, this story restores focus back to the trigger button instead of the referenced input element but specifically for the dialogs beyond the first (aka when you open multiple dialogs). useRestoreFocus seems to call restoreElement on the external input element properly, but it seems like something else focuses the nested dialog trigger button afterwards...

File: packages/@react-types/dialog/src/index.d.ts [hunk 4]
```diff
@@ -70,11 +70,17 @@ export interface SpectrumDialogProps extends AriaDialogProps, StyleProps {
...
+   * Whether to restore focus back to the element that was focused
+   * when the focus scope mounted, after the focus scope unmounts.
+   * Optionally accepts a RefObject for the element that should receive focus.
+   */
+  restoreFocus?: boolean | RefObject<DOMRefValue<HTMLElement>> | RefObject<HTMLElement>
 }
 
 export interface SpectrumAlertDialogProps extends DOMProps, StyleProps {
-  /** The [visual style](https://spectrum.adobe.com/page/alert-dialog/#Options) of the AlertDialog.  */
+  /** The [visual style](https://spectrum.adobe.com/page/dialog/#Options) of the AlertDialog.  */
```
Comment (LFDanLu) [inline packages/@react-types/dialog/src/index.d.ts:83]:
```suggestion
  /** The [visual style](https://spectrum.adobe.com/page/alert-dialog/#Options) of the AlertDialog.  */
```

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx (hunk 3 above)
Comment (majornista) [reply packages/@react-aria/focus/stories/FocusScope.stories.tsx:131]: @LFDanLu The `onFocus` event handler on the document forces focus back to the `activeScope` at:
https://github.com/adobe/react-spectrum/blob/19ada7736dec4687b91d3ca758a0570e1215c857/packages/%40react-aria/focus/src/FocusScope.tsx#L275-L281. This kind of makes sense. If you have a FocusScope that contains focus open, like a modal dialog, you don't want a user be able to force focus to some element underneath it.

Comment (LFDanLu) [reply packages/@react-aria/focus/stories/FocusScope.stories.tsx:131]: I agree, I think that prioritizing a restoreElement within top most active FocusScope over the user provided restoreElement ref makes sense. The thing that throws me off is that this [story](https://reactspectrum.blob.core.windows.net/reactspectrum/19ada7736dec4687b91d3ca758a0570e1215c857/storybook/index.html?path=/story/focusscope--restore-focus-ref) **always** moves focus to the referenced input element.
Reproduction steps:
1. go to https://reactspectrum.blob.core.windows.net/reactspectrum/19ada7736dec4687b91d3ca758a0570e1215c857/storybook/index.html?path=/story/focusscope--restore-focus-ref
2. Open multiple dialogs by clicking on the "open dialog" button for each sub dialog that opens
3. Close the "top" most dialog (aka the last one you opened) via the "close" button
4. Note that focus is moved to the "Restore focus here" input element even though there are still other open dialogs.
5. Try steps 1-3 again, but instead close the dialog via the escape key. Note that focus is to the preceeding dialog's trigger button (this is what I would expect should happen)

Contrast this vs the [portal story](https://reactspectrum.blob.core.windows.net/reactspectrum/19ada7736dec4687b91d3ca758a0570e1215c857/storybook/index.html?path=/story/focusscope--restore-focus-ref-inside-portal) which always restores focus to a preceding dialog's trigger button if there is one available.

Comment (LFDanLu) [reply packages/@react-aria/focus/stories/FocusScope.stories.tsx:131]: Just wanna bring up https://github.com/adobe/react-spectrum/pull/2445/files#r828533708 again, perhaps we won't run into this weirdness if we strip out support for a user provided restoreElement ref...

Comment (majornista) [reply packages/@react-aria/focus/stories/FocusScope.stories.tsx:131]: > The thing that throws me off is that this [story](https://reactspectrum.blob.core.windows.net/reactspectrum/19ada7736dec4687b91d3ca758a0570e1215c857/storybook/index.html?path=/story/focusscope--restore-focus-ref) **always** moves focus to the referenced input element.

Ok, I understand. It seems like we have a timing issue. With the portal we seem to execute the code I reference in https://github.com/adobe/react-spectrum/pull/2445#discussion_r833781088 after focus restores to the focus is restored to the restoreFocus ref.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 5]
```diff
@@ -31,6 +31,7 @@ interface FocusScopeProps {
   /**
    * Whether to restore focus back to the element that was focused
    * when the focus scope mounted, after the focus scope unmounts.
+   * Optionally accepts a RefObject for the element that should receive focus.
```
Comment (LFDanLu) [inline packages/@react-aria/focus/src/FocusScope.tsx:34]: We can go ahead and remove this line now

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx [hunk 6]
```diff
@@ -92,12 +94,12 @@ function Example({usePortal, contain}: StoryProps) {
...
   return (
     <div>
-      <input />
+      <input  />
 
       <button type="button" onClick={() => setOpen(true)}>
         Open dialog
       </button>
-      <input />
+
```
Comment (LFDanLu) [inline packages/@react-aria/focus/stories/FocusScope.stories.tsx:102]: Mind keeping this input here? It is helpful for testing the "contain=false" case, specifically for testing that tabbing forward through the dialog should land you on the element after the trigger

File: packages/@react-types/dialog/src/index.d.ts [hunk 7]
```diff
@@ -70,7 +70,12 @@ export interface SpectrumDialogProps extends AriaDialogProps, StyleProps {
   /** Whether the Dialog is dismissable. See the [examples](#examples) for more details. */
   isDismissable?: boolean,
   /** Handler that is called when the 'x' button of a dismissable Dialog is clicked. */
-  onDismiss?: () => void
+  onDismiss?: () => void,
+  /**
+   * Whether to restore focus back to the element that was focused
+   * when the focus scope mounted, after the focus scope unmounts.
+   */
+  restoreFocus?: boolean
```
Comment (LFDanLu) [inline packages/@react-types/dialog/src/index.d.ts:78]: Don't think we need this anymore

**Review by LFDanLu — COMMENTED** [review]
Just some small cleanup items, but I confirmed that the nodeToRestore behavior works properly.

**Review by LFDanLu — APPROVED** [review]
LGTM

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 8]
```diff
@@ -383,7 +383,23 @@ function useAutoFocus(scopeRef: RefObject<HTMLElement[]>, autoFocus: boolean) {
       }
     }
     autoFocusRef.current = false;
-  }, []);
+  }, [scopeRef]);
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:386]: when does scopeRef change that isn't on mount?

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 9]
```diff
@@ -458,12 +477,30 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
...
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        if (restoreFocus && isElementInScope(document.activeElement, scopeRef.current)) {
+          requestAnimationFrame(() => {
+            if (document.body.contains(nodeToRestore) && nodeToRestore !== document.body) {
+              focusElement(nodeToRestore);
+            } else {
+              while (index > 0) {
+                index--;
+                let node = nodeToRestoreArray[index];
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:491]: can there be items at indices greater than index? i think it'd be better if we actually treated nodeToRestoreArray as a stack and used `Array.prototype.pop()`
I wouldn't provide a method to remove certain indices either. I'd just pop, if it's not in the dom, continue popping.

otherwise it looks to me like there are some times where the nodeToRestore maybe isn't removed from the array? such as in the if statement on line 486?

File: packages/@react-aria/focus/src/useFocusable.tsx [hunk 10]
```diff
@@ -76,7 +76,7 @@ export function useFocusable(props: FocusableOptions, domRef: RefObject<HTMLElem
       domRef.current.focus();
     }
     autoFocusRef.current = false;
-  }, []);
+  }, [domRef]);
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/useFocusable.tsx:79]: why was this added? when does it change other than mount? or is this just lint cleanup?

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx [hunk 11]
```diff
@@ -61,6 +61,7 @@ function NestedDialog({onClose, usePortal, contain}: {onClose: VoidFunction, use
     <MaybePortal usePortal={usePortal}>
       <FocusScope contain={contain} restoreFocus autoFocus>
         {!showNew && (
+          // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
```
Comment (snowystinger) [inline packages/@react-aria/focus/stories/FocusScope.stories.tsx:64]: i assume this is just lint cleanup?

File: packages/@react-spectrum/dialog/src/Dialog.tsx [hunk 12]
```diff
@@ -71,7 +71,7 @@ function Dialog(props: SpectrumDialogProps, ref: DOMRef) {
     content: {UNSAFE_className: styles['spectrum-Dialog-content']},
     footer: {UNSAFE_className: styles['spectrum-Dialog-footer']},
     buttonGroup: {UNSAFE_className: classNames(styles, 'spectrum-Dialog-buttonGroup', {'spectrum-Dialog-buttonGroup--noFooter': !hasFooter}), align: 'end'}
-  }), [styles, hasFooter, hasHeader, titleProps]);
+  }), [hasFooter, hasHeader, titleProps]);
```
Comment (snowystinger) [inline packages/@react-spectrum/dialog/src/Dialog.tsx:74]: more lint cleanup? styles should just be the same object from initial loading of the module, or is it changing sometimes?

File: packages/@react-spectrum/dialog/test/DialogContainer.test.js [hunk 13]
```diff
@@ -27,7 +27,7 @@ describe('DialogContainer', function () {
   });
 
   beforeEach(() => {
-    jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => cb());
+    jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => setTimeout(cb, 0));
```
Comment (snowystinger) [inline packages/@react-spectrum/dialog/test/DialogContainer.test.js:30]: go ahead and change our start section for this file to 
```
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runAllTimers();
  });
```

I verified locally it'll work and it's once less incorrect implementation of mock request animation frame. Not your fault, we used to have to do that.

File: packages/@react-spectrum/dialog/test/DialogContainer.test.js [hunk 14]
```diff
@@ -43,6 +43,7 @@ describe('DialogContainer', function () {
     );
 
     let button = getByRole('button');
+    act(() => button.focus());
```
Comment (snowystinger) [inline packages/@react-spectrum/dialog/test/DialogContainer.test.js:46]: unneeded i'm pretty sure, why was it added?

File: packages/@react-spectrum/dialog/test/DialogContainer.test.js [hunk 15]
```diff
@@ -67,6 +67,7 @@ describe('DialogContainer', function () {
...
```
Comment (snowystinger) [inline packages/@react-spectrum/dialog/test/DialogContainer.test.js:70]: same question, triggering a press should move focus onto the button

File: packages/@react-spectrum/dialog/test/DialogContainer.test.js [hunk 16]
```diff
@@ -134,6 +136,7 @@ describe('DialogContainer', function () {
...
```
Comment (snowystinger) [inline packages/@react-spectrum/dialog/test/DialogContainer.test.js:139]: same with the rest of these

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 8 above)
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:386]: I don't think it does. This was added to clear the following ESLint warning:

> React Hook useEffect has a missing dependency: 'scopeRef'. Either include it or remove the dependency array.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)

File: packages/@react-aria/focus/src/useFocusable.tsx (hunk 10 above)
Comment (majornista) [reply packages/@react-aria/focus/src/useFocusable.tsx:79]: lint cleanup. I'll replace it with a comment.

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx (hunk 11 above)
Comment (majornista) [reply packages/@react-aria/focus/stories/FocusScope.stories.tsx:64]: Yes. The test is listening for `onKeyDown` on a non-interactive element.

File: packages/@react-spectrum/dialog/src/Dialog.tsx (hunk 12 above)
Comment (majornista) [reply packages/@react-spectrum/dialog/src/Dialog.tsx:74]: This was a lint cleanup. Should actually be:
`[hasFooter, hasHeader, hasHeading, hasTypeIcon, titleProps]`.

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 9 above)
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:491]: Yes there can. I think this is what happens when a dialog opens from a menu item and the menu then closes. The menu item gets added to the end of the nodeToRestoreArray, to serve as the node to restore for the dialog. Then the FocusScope for the menu unmounts removing the nodeToRestore for the Dialog from the DOM. We need to clean up the array, removing the node that is no longer present in the DOM. 

See this example: https://reactspectrum.blob.core.windows.net/reactspectrum/624fe15622aaea936834c922cb308877c83b3d83/storybook/index.html?path=/story/dialogcontainer--nested-dialog-containers

Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:491]: but when we're popping off items in the stack, won't we check if they are in the dom? so we don't need to clear them out ahead of time?

File: packages/@react-spectrum/dialog/test/DialogContainer.test.js (hunk 14 above)
Comment (majornista) [reply packages/@react-spectrum/dialog/test/DialogContainer.test.js:46]: I'm not so sure about that. `triggerPress` only seems to fire `mouseDown`, `mouseUp` and `click`, I don't think it forces focus. See 
https://github.com/adobe/react-spectrum/blob/418e59ba5be2418cb52ea5087162a6faeeb1cc5a/packages/dev/test-utils/src/events.ts#L19-L23

File: packages/@react-spectrum/dialog/src/Dialog.tsx (hunk 12 above)
Comment (majornista) [reply packages/@react-spectrum/dialog/src/Dialog.tsx]: `styles` isn't changing. It contains the classNames from spectrum.

File: packages/@react-spectrum/dialog/src/DialogTrigger.tsx [hunk 17]
```diff
@@ -63,7 +63,7 @@ function DialogTrigger(props: SpectrumDialogTriggerProps) {
         console.warn('A DialogTrigger unmounted while open. This is likely due to being placed within a trigger that unmounts or inside a conditional. Consider using a DialogContainer instead.');
       }
     };
-  }, []);
+  }, [type]);
```
Comment (snowystinger) [inline packages/@react-spectrum/dialog/src/DialogTrigger.tsx:66]: Is this in case it switches between types? is there a test case or storybook I should be looking at?

Comment (majornista) [reply packages/@react-spectrum/dialog/src/DialogTrigger.tsx:66]: This is once again for an ESLint warning.

**Comment by snowystinger** [comment]
Can we separate the actual work from all these unrelated lint fixes? It's making it hard to tell what to look at.
We have an open pr with lint fixes. I'll eventually get time to push more on that or break it into smaller pieces focused just on lint.

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 9 above)
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx]: I think I may now have a fix for this: https://github.com/adobe/react-spectrum/pull/2445/commits/a2fad80eaffb0724db6b247daefb0e118aea9a51

**Review by snowystinger — DISMISSED** [review]
Tested in Chrome and Safari, behaves well

**Comment by majornista** [comment]
@LFDanLu Can you take another look at this one?

**Review by LFDanLu — COMMENTED** [review]
I'll need to dig into it a some more but the following [story](https://reactspectrum.blob.core.windows.net/reactspectrum/046df7ffb08aff65b9c9db93b89c539be10f4efb/storybook/index.html?path=/story/focusscope--keyboard-navigation-no-contain&providerSwitcher-toastPosition=bottom) breaks in the following case:
1. Open the dialog with you keyboard
2. Open several layers of nested dialogs via the "open dialog" buttons in each new dialog
3. Alt + Tab back to the first "dialog" you opened and hit the "close" button with the enter key
4. Note that focus is lost instead of being restored to the first "open dialog" button

The above doesn't reproduce in an earlier version of this PR: https://reactspectrum.blob.core.windows.net/reactspectrum/c465126e8286623894aca7b692cb096121643211/storybook/index.html?path=/story/focusscope--keyboard-navigation-no-contain

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 18]
```diff
@@ -458,15 +466,28 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
...
-          }
-        });
+      if (nodeToRestore) {
+        if (restoreFocus && isElementInScope(document.activeElement, scopeRef.current)) {
+          requestAnimationFrame(() => {
+            // If the nodeToRestore is not the last in the nodeToRestoreArray,
+            // do nothing because a later focusScope is open and should take precedence.
+            if (nodeToRestoreArray.indexOf(nodeToRestore) !== nodeToRestoreArray.length - 1) {
+              return;
+            }
```
Comment (LFDanLu) [inline packages/@react-aria/focus/src/FocusScope.tsx:476]: Looks like this line is responsible for the behavior I noted here: https://github.com/adobe/react-spectrum/pull/2445#pullrequestreview-977644837. The above makes an assumption that the later FocusScope isn't affected by the unmounting of a preceding FocusScope right? Can we make that assumption?

Is this block of code an optimization or was there a case covered here?

**Review by LFDanLu — APPROVED** [review]
LGTM, confirmed the previously discovered issue is fixed. I was wondering if we would also need to check if the nodeToRestore was in the parent of the currently active scope but I couldn't really think of a realistic use case for that

**Comment by snowystinger** [comment]
Noticed while in testing session, can't tell if this is covered in a different PR of yours, but repro steps are

Modal story "unmounting trigger" I can open the popover via keyboard, then open the dialog, then without hitting tab or anything, I can re-open the popover by hitting Enter once more.

**Comment by majornista** [comment]
> Noticed while in testing session, can't tell if this is covered in a different PR of yours, but repro steps are
> 
> Modal story "unmounting trigger" I can open the popover via keyboard, then open the dialog, then without hitting tab or anything, I can re-open the popover by hitting Enter once more.

This is covered by #3128.

**Comment by majornista** [comment]
> > Noticed while in testing session, can't tell if this is covered in a different PR of yours, but repro steps are
> > Modal story "unmounting trigger" I can open the popover via keyboard, then open the dialog, then without hitting tab or anything, I can re-open the popover by hitting Enter once more.
> 
> This is covered as a separate issue under #3128. A suggested fix is offered there, but I want to wait on this PR before implementing it.

@snowystinger I opened a PR to address #3128. See #3171.

**Review by LFDanLu — APPROVED** [review]
LGTM, confirmed the DialogContainer stories and FocusScope stories are behaving as expected, focus is being returned to the proper node or a fallback node

File: packages/@react-spectrum/picker/test/Picker.test.js [hunk 19]
```diff
@@ -1289,6 +1289,7 @@ describe('Picker', function () {
       triggerPress(item3);
       expect(onSelectionChange).toHaveBeenCalledTimes(1);
       act(() => jest.runAllTimers());
+      act(() => jest.runAllTimers());
```
Comment (snowystinger) [inline packages/@react-spectrum/picker/test/Picker.test.js:1292]: note: this test used to be technically wrong, after the first runAllTimers here, focus was on the body, we never ran the queued restore focus which is queued up as a result of the render contained in the act.

Because we now have a stack of nodes to restore focus to, we need to make sure to run our cleanup in the raf to clear that stack out, otherwise we have extra elements in the stack that will cause problems further down in the test

it's the same for all other tests here where it's runalltimersx2

File: packages/@react-spectrum/overlays/stories/Modal.stories.tsx [hunk 20]
```diff
@@ -63,15 +63,15 @@ function UnmountingTrigger() {
     <Fragment>
       <DialogTrigger type="popover" isOpen={isPopoverOpen} onOpenChange={setPopoverOpen}>
         <ActionButton>Open popover</ActionButton>
-        <Dialog>
+        <Dialog isDismissable={false}>
```
Comment (snowystinger) [inline packages/@react-spectrum/overlays/stories/Modal.stories.tsx:66]: these should already be false? what are these changes for?

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 21]
```diff
@@ -460,15 +468,34 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
...
+            while (nodeToRestoreArray.length > 0) {
+              let node = nodeToRestoreArray.pop();
+              if (document.body.contains(node)) {
+                focusElement(node);
+                break;
+              }
+            }
+            return;
+          }
+          // Otherwise try to focus the nodeToRestore, and clean up
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:487]: i don't quite understand this case, can you explain when we'd hit this?

File: packages/@react-spectrum/overlays/stories/Modal.stories.tsx (hunk 20 above)
Comment (majornista) [reply packages/@react-spectrum/overlays/stories/Modal.stories.tsx:66]: The responsive Popover hides the footer containing the "Open modal" button when rendered as a Dialog which defaults to `isDismissible={true}`, making it difficult to test if your screen width is too narrow.

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 21 above)
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:487]: We have to account for FocusScopes where `contain={false}`, as @LFDanLu pointed out in this comment: https://github.com/adobe/react-spectrum/pull/2445#pullrequestreview-977644837

> I'll need to dig into it a some more but the following [story](https://reactspectrum.blob.core.windows.net/reactspectrum/046df7ffb08aff65b9c9db93b89c539be10f4efb/storybook/index.html?path=/story/focusscope--keyboard-navigation-no-contain&providerSwitcher-toastPosition=bottom) breaks in the following case:
>
>   1. Open the dialog with you keyboard
>   2. Open several layers of nested dialogs via the "open dialog" buttons in each new dialog
>   3. Alt + Tab back to the first "dialog" you opened and hit the "close" button with the enter key
>   4. Note that focus is lost instead of being restored to the first "open dialog" button
>
> The above doesn't reproduce in an earlier version of this PR: https://reactspectrum.blob.core.windows.net/reactspectrum/c465126e8286623894aca7b692cb096121643211/storybook/index.html?path=/story/focusscope--keyboard-navigation-no-contain
> 
> EDIT:
> Seems to be due to something in [a2fad80](https://github.com/adobe/react-spectrum/pull/2445/commits/a2fad80eaffb0724db6b247daefb0e118aea9a51)

Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:487]: We iterate backward through the `nodeToRestoreArray` using `pop()` only when the `nodeToRestore` for the closing scope is the last item in the array.

File: packages/@react-spectrum/picker/test/Picker.test.js (hunk 19 above)
Comment (majornista) [reply packages/@react-spectrum/picker/test/Picker.test.js:1292]: So to clarify, where I now have `act(() => {jest.runAllTimers();});` twice is appropriate for this case? I saw instances where `runAllTimers` was called twice, in `ActionBar.test` and `Picker.test`, and figured I'd give it a try to fix some failing tests here. Sure enough, it worked.

**Comment by snowystinger** [comment]
I looked into this with @devongovett , we came to the conclusion a stack, while close, isn't quite right. We want to try refactoring our Map `scopes` (which is currently a poor man's tree) into a real, doubly linked, tree. This way we can have sibling focus scopes and they won't try to restore to each other depending on the order they were added/removed. We also noticed that our Map isn't always up to date currently.

**Comment by majornista** [comment]
> I looked into this with @devongovett , we came to the conclusion a stack, while close, isn't quite right. We want to try refactoring our Map `scopes` (which is currently a poor man's tree) into a real, doubly linked, tree. This way we can have sibling focus scopes and they won't try to restore to each other depending on the order they were added/removed. We also noticed that our Map isn't always up to date currently.

> @devongovett and @snowystinger While I agree that refactoring the restoreFocus behavior in FocusScopes around the composite pattern is desirable, please consider that this PR has been open since October 2021, ~7 months. Provided it fixes the problem and does not introduce breaking changes, I think it would be better to accept the PR and address the more significant refactoring as a separate issue.

**Comment by devongovett** [comment]
Fixed by  #3323
